packages/next/src/lib/generate-interception-routes-rewrites.ts

```diff
@@ -32,30 +136,94 @@ export function generateInterceptionRoutesRewrites(
       const { interceptingRoute, interceptedRoute } =
         extractInterceptionRouteInformation(appPath)
 
-      const normalizedInterceptingRoute = `${
-        interceptingRoute !== '/' ? toPathToRegexpPath(interceptingRoute) : ''
-      }/(.*)?`
+      // Detect which marker is being used
+      const marker = getInterceptionMarker(appPath)
+
+      // The Next-Url header does not contain the base path, so just use the
+      // intercepting route.
+      const header = getNamedRouteRegex(interceptingRoute, {
+        prefixRouteKeys: true,
+      })
+
+      // The source is the intercepted route with the base path, it's matched by
+      // the router. Generate this first to get the correct parameter prefixes.
+      const source = getNamedRouteRegex(basePath + interceptedRoute, {
+        prefixRouteKeys: true,
+      })
+
+      // The destination should use the same parameter reference as the source
+      // so that parameter substitution works correctly. This ensures that when
+      // the router extracts params from the source, they can be substituted
+      // into the destination.
+      const destination = getNamedRouteRegex(basePath + appPath, {
+        prefixRouteKeys: true,
+        reference: source.reference,
+      })
+
+      // Generate the appropriate header regex based on the marker type
+      let headerRegex: string
+      if (marker === '(.)') {
+        // For same-level interception, match routes at the same level as the intercepting route
+        // Use header.reference which has the param -> prefixedKey mapping
+        headerRegex = generateSameLevelHeaderRegex(
+          interceptingRoute,
+          header.reference
+        )
+      } else if (marker === '(..)') {
+        // For parent-level interception, match routes at the intercepting route level
+        // Check if there's a catchall sibling at the intercepting route level
+        const hasCatchallSibling = hasCatchallSiblingAtLevel(
+          appPaths,
+          interceptingRoute
+        )
+
+        // Build regex pattern that handles dynamic segments correctly
+        const patterns: string[] = []
+        for (const segment of interceptingRoute.split('/').filter(Boolean)) {
+          const param = getSegmentParam(segment)
+          if (param) {
+            // Dynamic segment - use named capture group from header.reference
+            const key = header.reference[param.param]
+            if (!key) {
+              throw new InvariantError(
+                `No reference found for param: ${param.param} in reference: ${JSON.stringify(header.reference)}`
+              )
+            }
 
-      const normalizedInterceptedRoute = toPathToRegexpPath(interceptedRoute)
-      const normalizedAppPath = toPathToRegexpPath(appPath)
+            // Check if this is a catchall (repeat) parameter
+            if (isCatchAll(param.type)) {
+              patterns.push(`(?<${key}>.+?)`)
```

Comment:
does this properly handle optional catch-all? I would have expected a different regex for catch all vs optional catch-all since the non-specified route would match.